example/builder/src/new_class_consumed_props.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:over_react/over_react.dart';
+
+// ignore_for_file: uri_has_not_been_generated
+part 'new_class_consumed_props.over_react.g.dart';
+
+UiFactory<SomeParentProps> SomeClassParent = _$SomeClassParent; // ignore: undefined_identifier
+
+mixin ParentOnlyPropsMixin on UiProps {
+  String aParentProp;
+}
+
+mixin SharedPropsMixin on UiProps {
+  String aPropToBePassed;
+}
+
+class SomeParentProps = UiProps with ParentOnlyPropsMixin, SharedPropsMixin;
+
+class SomeClassParentComponent extends UiComponent2<SomeParentProps> {
+  @override
+  render() {
+    final meta = UiPropsMeta(props).meta;
+    meta.consume(ParentOnlyPropsMixin);
```

Comment:
Could this replace the original consumedProps mechanism? It would keep the API consistent and reduce some of what the builder needs to do, since this is available on all Component2 instances anyway.

The main tradeoff is that this system isn't as intuitive for class based components and could get unmanageable fast since they need to be re-consume the props every time a new props instance is received (which also has a performance implication, even if slight)

Reply:
(I'm commenting on the overall approach in this thread, not just specific to class components.)

I'm not crazy about the idea of storing the consumed props themselves on the props instance because:
- This is storing additional state on what's supposed to be purely a typed view into a map class.
- The state needs to be recreated each time it's used (e.g. in each render). This isn't huge deal, but it means that we wouldn't be to perform future optimizations like caching the full set of keys under the hood.
- This blurs the line of separation of concerns between storing the metadata for all props mixins used, and storing the set of props the consumer has opted into.

I would lean toward having the API be closer to what we have for `consumedProps`, which separates these concerns, and also provides consistency with the current syntax:
```dart
final consumedProps = props.meta.forMixins({ParentOnlyPropsMixin});
```


With that, you would also be able to memoize it if you really wanted to:
```dart
final _getConsumedProps = memoize2<Iterable<ConsumedProps>, PropsMetaCollection>((meta) => meta.forMixins({
  ParentOnlyPropsMixin,
}));

// or with a helper function that fills in type parameters:

final _getConsumedProps = memoizeConsumedProps((meta) => meta.forMixins({
  ParentOnlyPropsMixin,
}));

UiFactory<FooProps> Foo = uiFunction((props) {
  final consumedProps = _getConsumedProps(props.meta);
});
```

---

As far as using the consumed props goes, currently we have this:
```dart
..modifyProps((propsToChange) => meta.deriveUnconsumedProps(props, propsToChange))
```
I think we could add some helpers to help reduce this boilerplate.

Since we're passing through `propsToChange`, instead of creating that closure ourselves, we could instead return a function:
```dart
..modifyProps(meta.addUnconsumedPropsFrom(props))
```
Or even add something to UiProps:
```dart
..addUnconsumedProps(props, meta)
// Or if we end up using something like that `consumedProps` API I mentioned above
..addUnconsumedProps(props, consumedProps)
```

Reply:
Your APIs looked awesome - thanks for writing that up! Added those API changes in 9c7ebbe.